roundabout-core/src/main/java/at/fhv/itm3/s2/roundabout/event/CarCouldLeaveSectionEvent.java

```diff
@@ -1,18 +1,66 @@
+import desmoj.core.simulator.TimeSpan;
 
-public class CarCouldLeaveSectionEvent extends Event<Entity> implements ICarCouldLeaveSectionEvent {
+import java.util.concurrent.TimeUnit;
 
-    public CarCouldLeaveSectionEvent(Model model, String s, boolean b) {
-        super(model, s, b);
+public class CarCouldLeaveSectionEvent extends Event<StreetSection> {
+
+    /**
+     * A reference to the RoundaboutModel the CarCouldLeaveSectionEvent is part of
+     */
+    private RoundaboutModel myModel;
+
+    /**
+     * Constructs a new CarCouldLeaveSectionEvent
+     *
+     * @param model         the model this event belongs to
+     * @param name          this event's name
+     * @param showInTrace   flag to indicate if this event shall produce output for the trace
+     */
+    public CarCouldLeaveSectionEvent(Model model, String name, boolean showInTrace) {
+        super(model, name, showInTrace);
+        myModel = (RoundaboutModel)model;
     }
 
-    public void eventRoutine(Entity entity) throws SuspendExecution {
+    /**
+     * The event routine describes the moving of a car from one section to the next section.
+     *
+     * If the given section has a car at its exit point, it is checked if this car could leave this
+     * section and enter the next one. If that is true, a new CarCouldLeaveSectionEvent for the next
+     * section (the section the car enters) is scheduled after the time the car needs to traverse that
+     * section. After that the car is moved from this section to the next one.
+     * If the current section (the section in which the car was before moving) is not empty, a new
+     * CarCouldLeaveSectionEvent is scheduled for the current section after the transition time of the
+     * first car (only if this time is passed it makes sense to update all car positions on this time and
+     * check if there is another car to leave the section).
+     * Moreover for all previous sections of the current section a new CarCouldLeaveSectionEvent is immediately
+     * scheduled to check if the previous sections have a car which could enter the current section.
+     *
+     * @param section
+     * @throws SuspendExecution
+     */
+    public void eventRoutine(StreetSection section) throws SuspendExecution {
+        if(section.isFirstCarOnExitPoint()) {
+            if(section.carCouldEnterNextSection()) {
+                new CarCouldLeaveSectionEvent(myModel, "CarCouldLeaveSectionEvent", true).schedule((StreetSection)section.getFirstCar().getNextSection(), new TimeSpan(section.getFirstCar().getTimeToTraverseSection(section.getFirstCar().getNextSection()), TimeUnit.SECONDS));
+                section.moveFirstCarToNextSection();
+
+                if(!section.isEmpty()) {
+                    section.updateAllCarsPositions();
+                    new CarCouldLeaveSectionEvent(myModel, "CarCouldLeaveSectionEvent", true).schedule(section, new TimeSpan(section.getFirstCar().getTransitionTime(), TimeUnit.SECONDS));
```

Comment:
isn't this scheduled on CarGenerateEvent?

Reply:
in the CarGenerateEvent there is also a CarCouldLeaveSectionEvent scheduled, but here it is also necessary